docs/introduction.md

```diff
@@ -0,0 +1,282 @@
+The `List`, `Map`, and `Set` collections specialize _expressions_ in those domains
+both for performance and to facilitate unification and matching.
+
+References:
+
+- `Kore.Domain.Builtin`
+
+### `Condition`
+
+A `Condition` represents the conditions or constraints along an execution path,
+made up of a `Predicate` and a `Substitution`.
+
+References:
+
+- `Kore.Internal.Condition`
+
+#### `Predicate`
+
+`Predicate` is the representation predicate logic formulas,
+essentially the subset of `TermLike` that can be built from `Ceil` and connectives.
+`Predicate`s are the only type of pattern that can be translated for the external solver.
+
+References:
+
+- `Kore.Internal.Predicate`
+
+#### `Substitution`
+
+A `Substitution` is a collection of `(variable, TermLike variable)` pairs
+where `(x, t)` represents `\equals(x, t)`.
+`Substitution` can be viewed as a specialization of `Predicate` in that way,
+that is, a `Substitution` is the type of `Predicate` which can be applied as a substitution.
+
+References:
+
+- `Kore.Unification.Substitution`
+
+### `Pattern`
+
+A `Pattern` is part of a program configuration:
+some _term_ (`TermLike`) and the _condition_ on that term (`Condition`).
+It is unusual to operate on a term without the accompanying condition,
+so `Pattern` is somewhat more common than `TermLike` alone.
+
+References:
+
+- `Kore.Internal.Pattern`
+
+### `Conditional`
+
+`Conditional` is a generalization of `Pattern`:
+an arbitrary type with `Condition`s.
+
+References:
+
+- `Kore.Internal.Conditional`
+
+### `OrPattern`
+
+`OrPattern` is a disjunction of `Pattern`s (program configurations).
+This is the usual output of simplification.
+
+References:
+
+- `Kore.Internal.OrPattern`
+
+
+## Behavior
+
+### Unification
+
+Unification is represented in matching logic by `\and`
+and is essentially implemented as `\and` simplification,
+that is, by pushing `\and` down through symbols as far as possible.
+The backend has special support for unification modulo certain theories:
+
+- Constructors (injective, no-confusion)
+- Sort injections (constructors modulo triangle equality)
+- Overloaded symbols
+- `List` (common patterns only)
+- `Map` (common patterns only)
+- `Set` (common patterns only)
+
+Because unification is implemented as "pushing `\and` down", the solution is determined in parallel.
+Consider this example unification, with a constructor `C(_, _)` and constant terms `a` and `b` which may be undefined:
+
+```
+C(x, x) ∧ C(a, b)
+C(x ∧ a, x ∧ b)                                    -- constructor axioms: injectivity
+C(a ∧ (\ceil(a) ∧ x = a), b ∧ (\ceil(b) ∧ x = b))  -- singular variables
+C(a, b) ∧ (\ceil(a) ∧ \ceil(b)) ∧ (x = a ∧ x = b)  -- lifting conditions
+C(a, b) ∧ (\ceil(a) ∧ \ceil(b)) ∧ x = a ∧ b        -- substitution normalization
+C(a, b) ∧ (\ceil(a) ∧ \ceil(b)) ∧ x = (a ∧ a = b)  -- substitution normalization
+C(a, b) ∧ (\ceil(a) ∧ \ceil(b) ∧ a = b) ∧ x = a    -- substitution normalization
```

Comment:
As mentioned above, proper unification would that the `ceil` of this, see https://github.com/kframework/kore/blob/master/kore/src/Kore/Unification/Procedure.hs#L76